### PR TITLE
MPU9250 fixes and minor improvements

### PR DESCRIPTION
**Pull request by lucasdemarchi**
These are the fixes and minor improvements we have in order to enable MPU9250 on Minnowboard port.

### CODE REVIEW

**Comment by lucasdemarchi** [comment]
@enter-the-bot Doesn't seem to be the case here. We can talk to the AK8963 just fine. From the datasheet:

> To prevent switching into I2C mode when using SPI, the I2C interface should be disabled by setting the I2C_IF_DIS configuration bit. Setting this bit should be performed immediately after waiting for the time specified by the “Start-Up Time for Register Read/Write” in Section 6.3.
> For further information regarding the I2C_IF_DIS bit, please refer to the MPU-9250 Register Map and Register Descriptions document.

From the MPU-9250 Register Map:

> Rev. 1.4, page 40:
> 1 – Reset I2C Slave module and put the serial interface in SPI mode only.
> This bit auto clears after one clock cycle.

From the datasheet it seems you wouldn't be able to communicate with the AK8963 if you disabled the I2C master. In [AP_Compass_AK8963.cpp](../blob/master/libraries/AP_Compass/AP_Compass_AK8963.cpp) you can see the driver enables the slave0 and that the MPU9250 driver doesn't disable the I2C master functionality.

Am I missing anything here?

@guludo  ran the tests and told me the data from AK8963 is arriving just fine. @guludo any additional comment here?

**Comment by guludo** [comment]
Hi all,

So, as @lucasdemarchi pointed, I ran the tests yesterday. My test was to add the INS in [AP_Compass_test](https://github.com/diydrones/ardupilot/blob/master/libraries/AP_Compass/examples/AP_Compass_test/AP_Compass_test.cpp) in order to let both INS and compass running together.

MPU9250 has two I2C interfaces: one as slave and one as master. As far as we could verify, they seem to be independent.

**Comment by lucasdemarchi** [comment]
Just rebasing on master to make sure the patch set doesn't  bit rot.

File: libraries/AP_InertialSensor/AP_InertialSensor_MPU9250.cpp [hunk 1]
```diff
@@ -421,8 +417,14 @@ bool AP_InertialSensor_MPU9250::_hardware_init(void)
     // Chip reset
     uint8_t tries;
     for (tries = 0; tries<5; tries++) {
+        // reset device
+        _register_write(MPUREG_PWR_MGMT_1, BIT_PWR_MGMT_1_DEVICE_RESET);
+
         hal.scheduler->delay(100);
 
+        // disable I2C as recommended by the datasheet
+        _register_write(MPUREG_USER_CTRL, BIT_USER_CTRL_I2C_IF_DIS);
```
Comment (tridge) [inline libraries/AP_InertialSensor/AP_InertialSensor_MPU9250.cpp]: have you checked whether this impacts the use of the in-built AK8963 mag? (I seem to remember we had to have I2C enabled for that)

Comment (lucasdemarchi) [inline libraries/AP_InertialSensor/AP_InertialSensor_MPU9250.cpp]: Yes, we can read the AK8963 just fine.  However apparently there's a race between the initialization of INS and the Compass.... we are going to take a look there, but it seems orthogonal to this one.

**Comment by tridge** [comment]
applied, thanks!

**Comment by lucasdemarchi** [comment]
@enter-the-bot yeah... I'm not sure it's a race since the order in which the init() functions are called is deterministic. However AFAICS it's in the wrong order for copter. However it works fine for me the first 2~3 times and then after that it consistently fails the initialization. I'm working on it today.
